src/sql-parser/src/lexer.rs

```diff
@@ -0,0 +1,358 @@
+            ',' => Token::Comma,
+            '.' => Token::Dot,
+            ':' if buf.consume(':') => Token::DoubleColon,
+            ':' => Token::Colon,
+            ';' => Token::Semicolon,
+            '[' => Token::LBracket,
+            ']' => Token::RBracket,
+            #[rustfmt::skip]
+            '+'|'-'|'*'|'/'|'<'|'>'|'='|'~'|'!'|'@'|'#'|'%'|'^'|'&'|'|'|'`'|'?' => lex_op(buf)?,
+            _ => return Err(buf.err(pos, format!("unexpected character in input: {}", ch))),
+        };
+        tokens.push((token, pos..buf.pos()))
+    }
+    Ok(tokens)
+}
+
+fn lex_line_comment(buf: &mut LexBuf) {
+    buf.take_while(|ch| ch != '\n');
+}
+
+fn lex_multiline_comment(buf: &mut LexBuf) -> Result<(), ParserError> {
+    let pos = buf.pos() - 2;
+    let mut nesting = 0;
+    while let Some(ch) = buf.next() {
+        match ch {
+            '*' if buf.consume('/') => {
+                if nesting == 0 {
+                    return Ok(());
+                } else {
+                    nesting -= 1;
+                }
+            }
+            '/' if buf.consume('*') => nesting += 1,
+            _ => (),
+        }
+    }
+    Err(buf.err(pos, "unterminated multiline comment"))
+}
+
+fn lex_ident(buf: &mut LexBuf) -> Token {
+    buf.prev();
+    let word = buf.take_while(
+        |ch| matches!(ch, 'A'..='Z' | 'a'..='z' | '\u{80}'..='\u{ff}' | '0'..='9' | '$' | '_'),
+    );
+    let word_uppercase = word.to_uppercase();
+    Token::Word(Word {
+        value: word,
+        quoted: false,
+        keyword: if ALL_KEYWORDS.contains(&word_uppercase.as_str()) {
+            word_uppercase
+        } else {
+            "".into()
+        },
+    })
+}
+
+fn lex_quoted_ident(buf: &mut LexBuf) -> Result<Token, ParserError> {
+    let mut s = String::new();
+    let pos = buf.pos() - 1;
+    loop {
+        match buf.next() {
+            Some('"') if buf.consume('"') => s.push('"'),
+            Some('"') => break,
+            Some(c) => s.push(c),
+            None => return Err(buf.err(pos, "unterminated quoted identifier")),
+        }
+    }
+    Ok(Token::Word(Word {
+        value: s,
+        quoted: true,
+        keyword: "".into(),
+    }))
+}
+
+fn lex_string(buf: &mut LexBuf) -> Result<String, ParserError> {
+    let mut s = String::new();
+    loop {
+        let pos = buf.pos() - 1;
+        loop {
+            match buf.next() {
+                Some('\'') if buf.consume('\'') => s.push('\''),
+                Some('\'') => break,
+                Some(c) => s.push(c),
+                None => return Err(buf.err(pos, "unterminated quoted string")),
+            }
+        }
+        // Concatenate string literals that are separated by whitespace if and
+        // only if that whitespace contains at least one newline character. This
+        // bizzare rule matches PostgreSQL and the SQL standard.
+        let whitespace = buf.take_while(|ch| matches!(ch, ' ' | '\t' | '\n' | '\r' | '\x0c'));
```

Comment:
Yeah, I went back and forth on this a lot. On the one hand, it would be nice to reuse the definition of whitespace; on the other, it's nice to get exhaustive matching and better readability in the main `match` statement. One day, when patterns can include embedded `|`s, we'll be able to write a `whitespace!` macro that expands out to the whitespace character pattern, and then we'll have the best of both worlds.

I'll switch it to `is_ascii_whitespace` for now.